### PR TITLE
Discord Rich Presence support

### PR DESCRIPTION
**Pull request by znvjder**
Based on fork #2577 

I decided to copy the code from the base fork and rework it due to compilation problems.
What has changed:
- Most features tweaked, class rebuilt a bit. 
- Added enable/disable from client menu for Discord rich presence sync - disabled by default. 
- Any server owner who wants their own logo must create their own application (we need to create a tutorial on how to do this).
- Added a feature to manage from the server, but on the client side.
- Added the ability to display buttons to the discord-rpc code.
- Support emojis

Below are some of the features:
```lua
setDiscordApplicationID(string applicationid)
setDiscordRichPresenceDetails(string details)
setDiscordRichPresenceState(string state)
setDiscordRichPresenceAsset(string assetImage, string text)
setDiscordRichPresenceSmallAsset(string assetImage, string text)
setDiscordRichPresenceButton(number index, string text, string url)
resetDiscordRichPresenceData()
isDiscordRichPresenceConnected()
```

Example on Discord:
drpc

All that was left was to add a default resource to show additional data.
I have no idea what else to add, so I'm open to suggestions.

### CODE REVIEW

**Comment by CrosRoad95** [comment]
here's our favourite issue back again, vendor libraries inflate pull requests, either: 1. download sources from some server, 2. use submodule
because now reviewing it is hard due to laggy browser :)

**Comment by znvjder** [comment]
> here's our favourite issue back again, vendor libraries inflate pull requests, either: 1. download sources from some server, 2. use submodule
> 
> because now reviewing it is hard due to laggy browser :)

Sorry, but due to the fact that I have an edited version of discord-rpc I had to add it.

If you want i can redo it.

**Comment by CrosRoad95** [comment]
i don't want to discuss the same issue over and over for every case, i want general solution

**Comment by Dutchman101** [comment]
@CrosRoad95 Please leave comments like that to the project maintainers (MTA team/write access). Avoid dragging contributors down on their PR's, if those responsible want to update coding guidelines then they can do it separately and in its own discussion. Don't reply to this post, let's keep it on topic, thanks.

@znvjder Don't mind it, vendor\discord-rpc in this PR is fine

Who uses a mobile browser to review code anyways, im on desktop and it scrolls smoothly.
Nice PR, and heavily awaited!

**Comment by Fernando-A-Rocha** [comment]
Commenting for support and exposure. Nice PR, thank you!

File: Client/core/CDiscordRichPresence.h [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    void SetAsset(const char* szAsset, const char* szAssetText, bool bIsLarge = false);
+    void SetAssetLargeData(const char* szAsset, const char* szAssetText);
+    void SetAssetSmallData(const char* szAsset, const char* szAssetText);
+
+    bool ResetDiscordData();
+    bool SetPresenceState(const char* szState, bool bCustom = false);
+    bool SetPresenceDetails(const char* szDetails, bool bCustom = false);
+    bool SetPresenceButtons(unsigned short int iIndex, const char* szName, const char* szUrl);
+    bool SetDiscordRPCEnabled(bool bEnabled);
+    bool IsDiscordRPCEnabled();
```
Comment (Lpsd) [inline Client/core/CDiscordRichPresence.h]: `IsDiscordRPCEnabled` can be made `const`

File: Client/sdk/core/CDiscordInterface.h [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    virtual bool ResetDiscordData() = 0;
+    virtual bool SetPresenceState(const char* szState, bool bCustom) = 0;
+    virtual void SetAssetLargeData(const char* szAsset, const char* szAssetText) = 0;
+    virtual void SetAssetSmallData(const char* szAsset, const char* szAssetText) = 0;
+    virtual void SetPresenceStartTimestamp(const unsigned long ulStart) = 0;
+    virtual bool SetPresenceButtons(unsigned short int iIndex, const char* szName, const char* szUrl) = 0;
+    //virtual void SetPresenceEndTimestamp(const unsigned long ulEnd) = 0;
+
+    virtual bool SetDiscordRPCEnabled(bool bEnabled = false) = 0;
+    virtual bool IsDiscordRPCEnabled() = 0;
```
Comment (Lpsd) [inline Client/sdk/core/CDiscordInterface.h]: remember to update `const` here

**Review by Lpsd — COMMENTED** [review]
Mostly seems good to me, just a couple of small comments.

You also mentioned "due to the fact that I have an edited version of discord-rpc I had to add it." - in this case we would want to commit those changes to https://github.com/multitheftauto/discord-rpc and restore the `install_discord` script. You should be able to make a pull request to the `master` branch of that repository with the changes.

This way we retain the `install_discord` premake script & exclude the vendor files from the `mtasa-blue` repo, which is definitely preferable.

It would be really nice to have a test resource showing examples of all functions. They will need to be written for the wiki anyway.

Good work and thanks for picking up the branch!

File: Client/mods/deathmatch/logic/luadefs/CLuaDiscordDefs.cpp [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+        return false;
+
+    return true;
+}
+
+bool CLuaDiscordDefs::SetAppID(std::string strAppID)
+{
+    int appIDLength = strAppID.length();
+
+    if (appIDLength > 32)
```
Comment (Lpsd) [inline Client/mods/deathmatch/logic/luadefs/CLuaDiscordDefs.cpp]: Need to check for `appIDLength < 1` here

Comment (Allerek) [reply Client/mods/deathmatch/logic/luadefs/CLuaDiscordDefs.cpp]: > Need to check for `appIDLength < 1` here

Discord APPids are at least 17 characters long

Comment (TracerDS) [reply Client/mods/deathmatch/logic/luadefs/CLuaDiscordDefs.cpp]: > > Need to check for `appIDLength < 1` here
> 
> Discord APPids are at least 17 characters long

That's exactly why you need to check if its `< 17` or empty (`< 1`)

Comment (Lpsd) [reply Client/mods/deathmatch/logic/luadefs/CLuaDiscordDefs.cpp]: Either way, the expression needs to match the invalid argument exception message. Either option is fine but I would prefer not to be bound to limits which Discord could change, so a minimum of `1` character is preferable here.

Comment (znvjder) [reply Client/mods/deathmatch/logic/luadefs/CLuaDiscordDefs.cpp]: As for the number of characters in the application ID, it varies, sometimes it is 17, 18 or even 20. So I would stick with 32.

File: Client/sdk/core/CDiscordInterface.h (hunk 2 above)
Comment (znvjder) [reply Client/sdk/core/CDiscordInterface.h]: Done, sure we can do that, I don't see a problem. I will do it in a moment.

**Review by Lpsd — DISMISSED** [review]

**Comment by Lpsd** [comment]
To be merged along with this: https://github.com/multitheftauto/discord-rpc/pull/1

**Comment by Lpsd** [comment]
The branch got borked by our simultaneous commits, it's fixed now. I've kept the `v` in the tag for consistency (I made a mistake when creating `3.4.2`).

File: Client/core/CDiscordRichPresence.cpp [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+
+#include "StdInc.h"
+#include "discord_rpc.h"
+#include "CDiscordRichPresence.h"
+
+constexpr char DEFAULT_APP_ID[] = "1139666477813866546";
+constexpr char DEFAULT_APP_ASSET[] = "mta_logo_round";
+constexpr char DEFAULT_APP_ASSET_TEXT[] = "Multi Theft Auto";
+constexpr char DEFAULT_APP_ASSET_SMALL[] = "";
+constexpr char DEFAULT_APP_ASSET_SMALL_TEXT[] = "";
```
Comment (AlexTMjugador) [inline Client/core/CDiscordRichPresence.cpp]: Nice work and PR, this feature has been awaited for a long time! :heart:

As far as I can see, scripts can customize rich presence information without changing the app ID, which in the eyes of Discord would establish a link between MTA and whatever server owners decide to use rich presence for. Should we force scripts to use their own app ID if they want to change the default rich presence data?

Also, out of curiosity, who manages the default app ID? Could it get unexpectedly deleted or misused by its owner?

Comment (znvjder) [reply Client/core/CDiscordRichPresence.cpp]: I aimed to enable the scripts to establish their own application, hence create their image (emblem, logo?) in Rich Presence and other essential information.

Comment (AlexTMjugador) [reply Client/core/CDiscordRichPresence.cpp]: Right. It is my understanding that application IDs should not be treated as private information, and any third party can try to extract a rich presence application ID and use it for a completely unrelated application, so these concerns probably don't mean much from a security standpoint. But I still find them interesting.

Comment (Lpsd) [reply Client/core/CDiscordRichPresence.cpp]: > Should we force scripts to use their own app ID if they want to change the default rich presence data?

Good thinking. We should stop customization unless the server is using their own application ID.

> Also, out of curiosity, who manages the default app ID?

I think that's @znvjder's app ID. We'll get one created under the MTA discord, nice spot.

Comment (Lpsd) [reply Client/core/CDiscordRichPresence.cpp]: Please address the first comment above @znvjder  - unless the server has set their own application ID they shouldn't be able to customize anything.

Comment (znvjder) [reply Client/core/CDiscordRichPresence.cpp]: The given ID in the default application was for the time of testing, sorry for that, so MTA team has to give their own ID.

Generally there is a variable to lock it, just need to add some conditions to the functions. But in general - do we need to block all features if we don't want to use a custom application?

Comment (Lpsd) [reply Client/core/CDiscordRichPresence.cpp]: Yeah I'd say block everything

Comment (Allerek) [reply Client/core/CDiscordRichPresence.cpp]: > The given ID in the default application was for the time of testing, sorry for that, so MTA team has to give their own ID.
> 
> Generally there is a variable to lock it, just need to add some conditions to the functions. But in general - do we need to block all features if we don't want to use a custom application?

I think every server should set-up their own application id.

Comment (znvjder) [reply Client/core/CDiscordRichPresence.cpp]: Already all features are blocked till we set up own application. :)

**Review by Lpsd — APPROVED** [review]

**Pull request merged**